rfc/rfc-0000-blueprint-match-fields.md

```diff
@@ -0,0 +1,110 @@
+# Meta
+[meta]: #meta
+- Name: Add MatchFields to top level of blueprints
+- Start Date: 2022-02-10
+- Author(s): @jwntrs
+- Status: Draft
+- RFC Pull Request:
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+The [template switching RFC](https://github.com/vmware-tanzu/cartographer/pull/75) introduced template `options` that included a `matchFields` selector. We should include this same selector at the `blueprint` level
+
+# Motivation
+[motivation]: #motivation
+
+- Allow entire `blueprints` to match an `owner` based on certain fields.
+
+- Ensure that the behaviour of all `selectors` is consistent.
+
+# What it is
+[what-it-is]: #what-it-is
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSupplyChain
+metadata:
+  name: supply-chain
+spec:
+  selector:
+    matchFields:                                                # <=========== add this
+      { key: "workload.spec.image", operation: exists }         # <=========== 
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The same way as `matchFields` from [template switching RFC](https://github.com/vmware-tanzu/cartographer/pull/75)
+
+This will also impact how best match works at the supply chain level. Right now a best-match is indicated by the number of labels. We should broaden this definition to include the number of labels and number of fields matched.
+
+# Migration
+[migration]: #migration
+
+This is most likely a breaking change which would require a bump to the `ClusterSupplyChain` `apiVersion` and introducing a conversion webhook.
+
+```yaml
+  selector:
+    app.tanzu.vmware.com/workload-type: web
+```
+
+needs to become:
+
+```yaml
+  selector:
+    matchFields:
+      - key: workload.metadata.labels."app.tanzu.vmware.com/workload-type"
+        operator: In
+        values: ["web"]
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is a breaking change.
+
+# Alternatives
+[alternatives]: #alternatives
+
+If we introduced `matchLabels` at the same time as `matchFields`, when migrating we would simply need to hoist everything under the current `selector` over to the new `matchLabels` field.
```

Comment:
I would prefer having both `matchFields` and `matchLabels`, rather than forcing users to refer to labels through the `matchFields` key.

There are existing k8s resources (like Deployment) that use `selector.matchLabels` and I would want to keep that convention.

Also, using `matchFields` for labels would be more painful because you'd have to define the operator as well, rather than just using regular regular selection.

Reply:
updated the RFC to reflect adding both top level selectors